app/assets/src/components/views/UserPreferencesView.jsx

```diff
@@ -0,0 +1,134 @@
+import React from "react";
+import { set } from "lodash/fp";
+
+import { ViewHeader, NarrowContainer } from "~/components/layout";
+import {
+  getUserPreferenceMetadataByCategory,
+  updateUserPreference,
+} from "~/api";
+import LoadingMessage from "~/components/common/LoadingMessage";
+import Checkbox from "~ui/controls/Checkbox";
+import { UserContext } from "~/components/common/UserContext";
+
+import cs from "./user_preferences_view.scss";
+
+export default class UserPreferencesView extends React.Component {
+  state = {
+    userPreferenceMetadata: null,
+    // User preference values that were modified on this page and not reflected in the UserContext.
+    modifiedUserPreferences: {},
+    error: "",
+  };
+
+  async componentDidMount() {
+    const userPreferenceMetadata = await getUserPreferenceMetadataByCategory();
+
+    this.setState({
+      userPreferenceMetadata,
+    });
+  }
+
+  getUserPreferenceValues = () => {
+    const { userPreferences } = this.context || {};
+
+    return {
+      ...userPreferences,
+      ...this.state.modifiedUserPreferences,
+    };
+  };
+
+  handleUserPreferenceUpdate = async (key, value) => {
+    // Modify the value on the front-end, so that the user gets instant feedback.
+    this.setState({
+      modifiedUserPreferences: set(
+        key,
+        value,
+        this.state.modifiedUserPreferences
+      ),
+    });
+
+    try {
+      await updateUserPreference(key, value);
+    } catch {
+      // Show an error if the update fails.
```

Comment:
revert the state?

Reply:
That's a great idea, but I'd like to defer it because trying to revert state asynchronously introduces the possibility for race conditions and I'd like to keep the logic simple if possible. I think it's enough to have an error message to alert the user that something is wrong.